### PR TITLE
use interfaces for core contracts

### PR DESCRIPTION
**Pull request by jaycenhorton**
Temporarily setting this aside as it's turning out to be high-effort low-impact. 

- [all about interfaces](https://medium.com/coinmonks/solidity-tutorial-all-about-interfaces-f547d2869499)
- the interface was initially generated using `cast interface artifacts/contracts/Removal.sol/Removal.json -o contracts/interfaces/IRemoval.sol -n IRemoval` Which defines the interface really well, but duplicates inherited interfaces. The commented out sections in IRemoval are all inherited and thus are duplicates. The only thing I manually added was the errors and the interface inheritance and the names of the struct keys which weren't auto detected.

Opening this PR now though as the interfaces generated provide a really good view into what functions are public even if they are inherited which allows us to more easily guarantee we've appropriately constrained access to these functions as well as added pausability.

### CODE REVIEW

File: contracts/interfaces/ILockedNORIV2 [hunk 1]
```diff
@@ -0,0 +1,73 @@
+pragma solidity ^0.8.10;
+
+interface ILockedNORIV2 {
```
Comment (scottynomad) [inline contracts/interfaces/ILockedNORIV2:3]: Looks like a bug that this in missing an file extension.

File: contracts/Removal.sol [hunk 2]
```diff
@@ -361,7 +340,7 @@ contract Removal is
     uint256[] memory ids,
     uint256[] memory amounts,
     bytes memory data
-  ) public override {
+  ) public override(IERC1155Upgradeable, ERC1155Upgradeable) {
```
Comment (scottynomad) [inline contracts/Removal.sol:343]: why are both the interface and the implementation contract in here?

File: contracts/Removal.sol [hunk 3]
```diff
@@ -385,7 +364,11 @@ contract Removal is
   function supportsInterface(bytes4 interfaceId)
     public
     view
-    override(ERC1155Upgradeable, AccessControlEnumerableUpgradeable)
+    override(
+      IERC165Upgradeable,
```
Comment (scottynomad) [inline contracts/Removal.sol:368]: Is change related?

**Review by scottynomad — COMMENTED** [review]
What problem are we solving here?  These are interesting for their documentation value but don't seem to change the code significantly.

What would be the ongoing dev workflow here?  Are these treated like pure codegen that gets rewritten as part of the build process or do people need to update them manually?

File: contracts/Removal.sol (hunk 2 above)
Comment (jaycenhorton) [reply contracts/Removal.sol:343]: Was required

File: contracts/interfaces/ILockedNORIV2 (hunk 1 above)
Comment (jaycenhorton) [reply contracts/interfaces/ILockedNORIV2:3]: Yeah quickly generated these. So probably just a bug on my cli use

**Comment by jaycenhorton** [comment]
> What problem are we solving here? These are interesting for their documentation value but don't seem to change the code significantly.
> 
> What would be the ongoing dev workflow here? Are these treated like pure codegen that gets rewritten as part of the build process or do people need to update them manually?

As I mentioned I identified this as low impact, high effort, so this is likely to not be completed.


I haven't seen a **great** reason to prefer interfaces over importing the contracts themselves yet, but there are lots of points linked in the article in the original comment that are noteworthy 

**The primary reason I generated these was for reference, anything beyond that is less interesting**


However, I was also interested in using interfaces as a means to find:
- some kind of consistency for error, event and struct declarations that didn't require importing the contract code into files that shared these definitions. I think the net result is a decreased contract size which our solc runs constantly warn that we are over limit for. 
- the only way to use the inheritdoc natspec tag is via interfaces
- defining events in interfaces seems like the only clean way to re-use event definitions in forge without having to redefine them

Afaik there's no codegen process for declaring interfaces. They are usual

**Pull request closed without merging**